File: 2/2.py
## STAR 1
def star1():
    with open("input.txt", "r") as f:
        h = 0
        d = 0
        data = f.readlines()
        for x in data:
            splitted = x.split(" ")
            if splitted[0] == "forward":
                h += int(splitted[1])
            elif splitted[0] == "down":
                d += int(splitted[1])
            elif splitted[0] == "up":
                d -= int(splitted[1])
        return h * d


## STAR 2
def star2():
    with open("input.txt", "r") as f:
        h = 0
        d = 0
        a = 0
        data = f.readlines()
        for x in data:
            splitted = x.split(" ")
            if splitted[0] == "forward" and a != 0:
                h += int(splitted[1])
                d += int(splitted[1]) * a
            if splitted[0] == "forward" and a == 0:
                h += int(splitted[1])
            elif splitted[0] == "down":
                a += int(splitted[1])
            elif splitted[0] == "up":
                a -= int(splitted[1])
        return h * d
            

def main():
    print("Star 1: ", star1())
    print("Star 2: ", star2())

if __name__ == '__main__':
    main()

File: 2/2.hs
import Data.List.Split

t1 :: [String] -> Int -> Int -> Int 
t1 [] d h     = d*h
t1 (x:xs) d h | command == "forward" = t1 xs d (h + number)
              | command == "up"      = t1 xs (d - number) h
              | command == "down"    = t1 xs (d + number) h
              where
                  [command, number_string] = splitOn " " x
                  number = read number_string::Int

t2 :: [String] -> Int -> Int -> Int -> Int
t2 [] d h a   = d*h
t2 (x:xs) d h a | command == "forward" = t2 xs (d + number*a) (h + number) a
                | command == "up"      = t2 xs d h (a - number)
                | command == "down"    = t2 xs d h (a + number)
                where
                  [command, number_string] = splitOn " " x
                  number = read number_string::Int

main :: IO()
main = do
    input <- readFile "input.txt"
    let x = lines input
    print $ t1 x 0 0 
    print $ t2 x 0 0 0